pkg/flow/internal/controller/node_config_import.go

```diff
@@ -0,0 +1,328 @@
+			switch fullName {
+			case "declare":
+				cn.processDeclareBlock(stmt)
+			case importsource.BlockImportFile, importsource.BlockImportString: // TODO: add other import sources
+				cn.processImportBlock(stmt, fullName)
+			default:
+				level.Error(cn.logger).Log("msg", "only declare and import blocks are allowed in a module", "forbidden", fullName)
+			}
+		default:
+			level.Error(cn.logger).Log("msg", "only declare and import blocks are allowed in a module")
+		}
+	}
+}
+
+// processDeclareBlock stores the declare definition in the importedDeclares.
+func (cn *ImportConfigNode) processDeclareBlock(stmt *ast.BlockStmt) {
+	if _, ok := cn.importedDeclares[stmt.Label]; ok {
+		level.Error(cn.logger).Log("msg", "declare block redefined", "name", stmt.Label)
+		return
+	}
+	cn.importedDeclares[stmt.Label] = stmt.Body
+}
+
+// processDeclareBlock creates an ImportConfigNode child from the provided import block.
+func (cn *ImportConfigNode) processImportBlock(stmt *ast.BlockStmt, fullName string) {
+	sourceType := importsource.GetSourceType(fullName)
+	if _, ok := cn.importConfigNodesChildren[stmt.Label]; ok {
+		level.Error(cn.logger).Log("msg", "import block redefined", "name", stmt.Label)
+		return
+	}
+	childGlobals := cn.globals
+	// Children have a special OnBlockNodeUpdate function which notifies the parent when its content changes.
+	childGlobals.OnBlockNodeUpdate = cn.onChildrenContentUpdate
+	cn.importConfigNodesChildren[stmt.Label] = NewImportConfigNode(stmt, childGlobals, sourceType)
+}
+
+// evaluateChildren evaluates the import nodes managed by this import node.
+func (cn *ImportConfigNode) evaluateChildren() error {
+	for _, child := range cn.importConfigNodesChildren {
+		err := child.Evaluate(&vm.Scope{
+			Parent:    nil,
+			Variables: make(map[string]interface{}),
+		})
+		if err != nil {
+			return fmt.Errorf("imported node %s failed to evaluate, %v", child.label, err)
+		}
+	}
+	return nil
+}
+
+// onChildrenContentUpdate notifies the parent that the content has been updated.
+func (cn *ImportConfigNode) onChildrenContentUpdate(child BlockNode) {
+	// If the node is already updating its content, it will call OnBlockNodeUpdate
+	// so the notification can be ignored.
+	if !cn.inContentUpdate.Load() {
+		cn.OnBlockNodeUpdate(cn)
+	}
+}
+
+// Run runs the managed source and the import children until ctx is
+// canceled. Evaluate must have been called at least once without returning an
+// error before calling Run.
+//
+// Run will immediately return ErrUnevaluated if Evaluate has never been called
+// successfully. Otherwise, Run will return nil.
+func (cn *ImportConfigNode) Run(ctx context.Context) error {
+	cn.mut.Lock()
+	importChildrenCount := len(cn.importConfigNodesChildren)
+	cn.mut.Unlock()
+	if cn.source == nil {
+		return ErrUnevaluated
+	}
+
+	newCtx, cancel := context.WithCancel(ctx)
+	defer cancel() // This will stop the children and the managed source.
+
+	errChan := make(chan error, 1)
```

Comment:
tip: use [run.Group](https://pkg.go.dev/github.com/oklog/run#Group) here to launch multiple goroutines which can fail; there's examples of this being done elsewhere in our code.